e2e/test/iothub/SasCredentialAuthenticationTests.cs

```diff
@@ -0,0 +1,225 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Common.Exceptions;
+using Microsoft.Azure.Devices.E2ETests.Helpers;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+#if !NET451
+
+using Microsoft.Rest;
+using Azure;
+
+#endif
+
+using ClientOptions = Microsoft.Azure.Devices.Client.ClientOptions;
+
+namespace Microsoft.Azure.Devices.E2ETests.iothub
+{
+    /// <summary>
+    /// Tests to ensure authentication using SAS credential succeeds in all the clients.
+    /// </summary>
+    [TestClass]
+    [TestCategory("E2E")]
+    [TestCategory("IoTHub")]
+    public class SasCredentialAuthenticationTests : E2EMsTestBase
+    {
+        private readonly string _devicePrefix = $"E2E_{nameof(SasCredentialAuthenticationTests)}_";
+
+#if !NET451
+
+        [Ignore]
+        [LoggedTestMethod]
+        public async Task RegistryManager_Http_SasCredentialAuth_Success()
+        {
+            // arrange
+            string signature = Configuration.IoTHub.GetIotHubSharedAccessSignature(TimeSpan.FromHours(1));
+            using var registryManager = RegistryManager.Create(
+                Configuration.IoTHub.GetIotHubHostName(),
+                new AzureSasCredential(signature));
+
+            var device = new Device(Guid.NewGuid().ToString());
+
+            // act
+            Device createdDevice = await registryManager.AddDeviceAsync(device).ConfigureAwait(false);
+
+            // assert
+            Assert.IsNotNull(createdDevice);
+
+            // cleanup
+            await registryManager.RemoveDeviceAsync(device.Id).ConfigureAwait(false);
+        }
+
+        [Ignore]
+        [LoggedTestMethod]
+        public async Task RegistryManager_Http_SasCredentialAuth_Renewed_Success()
+        {
+            // arrange
+            string signature = Configuration.IoTHub.GetIotHubSharedAccessSignature(TimeSpan.FromHours(-1));
```

Comment:
This is probably me being paranoid, but I'd make the SAS token as expired as possible. We aren't trying to test how accurately the service evaluates when our SAS tokens expire. Maybe have them expire at the Unix Epoch rather than 1 hour in the past

Reply:
I think that would be really bad if the service does not evaluate it correctly. I think it is better we catch it and report to them rather than simply trying to make our tests pass.